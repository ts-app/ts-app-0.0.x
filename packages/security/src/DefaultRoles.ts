/**
 * Pre-defined roles used by MongoSecurityService.
 */
export enum DefaultRoles {
  /**
   * Administrator role.
   */
  Administrator = 'Administrator',
  /**
   * User role.
   */
  User = 'User'
}
