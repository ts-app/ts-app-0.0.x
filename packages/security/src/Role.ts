/**
 * This type specifies a name and unique ID for a Role.
 */
export interface Role {
  id: string
  name: string
}
