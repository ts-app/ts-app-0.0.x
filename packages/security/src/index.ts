export * from './DefaultRoles'
export * from './MongoRoleService'
export * from './MongoSecurityService'
export * from './Role'
export * from './RoleService'
export * from './securitySchemaDefinition'
export * from './SecurityService'
export * from './validators'
