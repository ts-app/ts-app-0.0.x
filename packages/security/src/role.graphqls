type Query {
    userIsInRoles(userId: String!, roles: [String]!, group: String): Boolean
    getGroupsForUser(userId: String, role: String): GetGroupsForUserPayload
    findRolesWithCursor(q: String, limit: Int, cursor: String): FindRolesWithCursorPayload
    userRoles(userId: String!, group: String): GetRolesForUserPayload
    getUsersInRoles(roles: [String]!, group: String, limit: Int, cursor: String): GetUsersInRolesPayload
}

type Mutation {
    addUsersToRoles(userIds: [String]!, roles: [String], group: String): NoPayload
    createRole(name: String!): CreateRolePayload
    removeRole(name: String!): NoPayload
    removeUsersFromRoles(userIds: [String]!, roles: [String]!, group: String): NoPayload
    removeUsersFromAllRoles(userIds: [String]!): NoPayload
}

type Role {
    id: String
    name: String
}

type CreateRolePayload {
    error: String
    id: String
}

type GetGroupsForUserPayload {
    error: String
    groups: [String]
}

type FindRolesWithCursorPayload {
    error: String
    cursor: String
    docs: [Role]
}

type GetRolesForUserPayload {
    error: String
    roles: [String]
}

type GetUsersInRolesPayload {
    error: String
    cursor: String
    docs: [User]
}
