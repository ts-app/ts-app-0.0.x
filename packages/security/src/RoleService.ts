import { Role } from './Role'
import { FindWithCursor } from '@ts-app/mongo'
import { User } from '@ts-app/common'

export const RoleServiceConstants = {
  GLOBAL: '_global_'
}

/**
 * This service allows assignment of roles to users for authorization purposes. It also provides
 * functions to check roles assigned to user, based on groups.
 *
 * Credits: This interface was heavily inspired by alanning/roles-npm.
 */
export interface RoleService {
  /**
   * Add users to specified roles, optionally specifying a group. If group name is not specified,
   * roles are assigned to RoleServiceConstants.GLOBAL group.
   */
  addUsersToRoles (input: {
    userIds: string | string[], roles: string | string[],
    group?: string
  }): Promise<{ error?: string }>

  /**
   * Create specified role name.
   */
  createRole (input: { name: string }): Promise<{ error?: string, id: string }>

  /**
   * Delete specified role name.
   *
   * If role is in use, service output error with appropriate message will be returned.
   */
  removeRole (input: { name: string }): Promise<{ error?: string }>

  /**
   * Get user's groups where user has role(s) assigned. RoleServiceConstants.GLOBAL will be omitted.
   *
   * If role is specified, restrict groups returned to those with specified role assigned.
   */
  getGroupsForUser (input: { userId: string, role?: string }): Promise<{ error?: string, groups?: string[] }>

  /**
   * Find roles with option to filter, limit and perform pagination on results.
   */
  findRolesWithCursor (input?: { q?: string, limit?: number, cursor?: string }): Promise<FindWithCursor<Role>>

  /**
   * Get roles assigned to user, optionally restricting specifying a group. Roles assigned to
   * RoleServiceConstants.GLOBAL will be included.
   *
   * If group name is not specified, all roles including those assigned to
   * RoleServiceConstants.GLOBAL are returned.
   */
  getRolesForUser (input: { userId: string, group?: string }): Promise<{ error?: string, roles?: string[] }>

  /**
   * Get users with specified role(s).
   */
  getUsersInRoles (input: { roles: string | string[], group?: string, limit?: number, cursor?: string }): Promise<FindWithCursor<User>>

  /**
   * Unassign user(s) from specified role(s).
   */
  removeUsersFromRoles (input: {
    userIds: string | string[], roles: string | string[],
    group?: string
  }): Promise<{ error?: string }>

  /**
   * Remove all assigned roles from specified user(s).
   */
  removeUsersFromAllRoles (input: { userIds: string | string[] }): Promise<{ error?: string }>

  /**
   * Returns true if user is assigned to one or more of the specified rule(s). This can be further
   * restricted based on group name.
   */
  userIsInRoles (input: { userId: string, roles: string | string[], group?: string }): Promise<boolean>
}
