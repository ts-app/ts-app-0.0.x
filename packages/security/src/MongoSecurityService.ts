import * as passport from 'passport'
import { StrategyOptions, Strategy, ExtractJwt } from 'passport-jwt'
import { Request, Response, NextFunction } from 'express'
import * as jwt from 'jsonwebtoken'
import * as bcryptjs from 'bcryptjs'
import { FindWithCursor, MongoService } from '@ts-app/mongo'
import { escapeRegex, Profile, User } from '@ts-app/common'
import { validateEmail, validatePassword } from './validators'
import { SecurityService } from './SecurityService'
import {
  Resolver,
  SecurityMiddleware,
  ServiceInfo,
  userIdMatchParam,
  environmentVariableIsSet
} from '@ts-app/graphql'
import { RoleService } from './RoleService'
import { DefaultRoles } from './DefaultRoles'

/**
 * SecurityService implementation that uses MongoDB as the data storage.
 *
 * This implementation also provides an Express middleware that configures JWT authentication.
 */
export class MongoSecurityService implements SecurityService, SecurityMiddleware, ServiceInfo {
  private mongoService: MongoService
  private roleService: RoleService
  private secretOrKey: string
  private accessTokenExpiresIn: string
  private refreshTokenExpiresIn: string

  constructor ({ mongoService, roleService, secretOrKey = 'taktau', accessTokenExpiresIn = '30m', refreshTokenExpiresIn = '30d' }: {
    mongoService: MongoService, roleService: RoleService
    secretOrKey?: string, accessTokenExpiresIn?: string, refreshTokenExpiresIn?: string
  }) {
    this.mongoService = mongoService
    this.roleService = roleService
    this.secretOrKey = secretOrKey
    this.accessTokenExpiresIn = accessTokenExpiresIn
    this.refreshTokenExpiresIn = refreshTokenExpiresIn

    this.configurePassport()
  }

  info (): { serviceName: string; } {
    return {
      serviceName: 'SecurityService'
    }
  }

  /**
   *
   * @param {string[]} roles List of roles authorized to access resource. Empty array allows any
   * authenticated request to access resource. Undefined/Null value allows authenticated requests
   * to access resource.
   * @return {(req: Request, res: Response, next: e.NextFunction) => any}
   */
  middleware (roles?: string[] | undefined) {
    const mustBeLoggedIn = !!roles
    const authorizedRoles = roles || []

    return (req: Request, res: Response, next: NextFunction) => {
      passport.authenticate('jwt', async (err: any, user: User | false, info: any) => {
        if (err) {
          console.error(err)
          res.status(500).send('Authentication error!')
        } else {
          if (user) {
            // --- request auth: step 2 of 3. assign user to request if authentication was successful
            req.user = user
            if (authorizedRoles && authorizedRoles.length > 0) {
              const isAuthorized = await this.roleService.userIsInRoles({
                userId: user.id,
                roles: authorizedRoles
              })
              if (isAuthorized) {
                // allow authenticated AND authorized access
                next()
              } else {
                // authenticated but not authorized (no access)
                res.status(401).send('Unauthorized access!')
              }
            } else {
              // allow any authenticated access
              next()
            }
          } else {
            // not authenticated
            if (mustBeLoggedIn) {
              res.status(401).send('Unauthenticated access!')
            } else {
              // allow unauthenticated access
              next()
            }
          }
        }
      })(req, res, next)
    }
  }

  configurePassport () {
    const loginWithUserId = this.loginWithUserId.bind(this)
    const jwtFromRequest = ExtractJwt.fromExtractors([
      ExtractJwt.fromAuthHeaderAsBearerToken(), ExtractJwt.fromUrlQueryParameter('accessToken')
    ])
    const strategyOptions: StrategyOptions = { jwtFromRequest, secretOrKey: this.secretOrKey }
    const strategy = new Strategy(strategyOptions, async (payload, next) => {
      try {
        // --- request auth: step 1 of 3. extract JWT token and perform user login if need
        const user = await loginWithUserId(payload.userId)
        if (!user.error) {
          // valid user login
          next(null, user.user)
        } else {
          // invalid user login
          next(null, false)
        }
      } catch (e) {
        console.error('Authentication error.', e)
        next(true)
      }
    })
    passport.use(strategy)
  }

  @Resolver({ type: 'mutation' })
  async signUp (input: { email: string, password: string }): Promise<{ error?: string, user?: User }> {
    const { email, password } = input

    if (!validateEmail(email)) {
      return { error: 'Invalid email' }
    }
    if (!validatePassword(password)) {
      return { error: 'Invalid password' }
    }

    const emailExist = await this.mongoService.get<User>('users', { 'emails.email': email })
    if (emailExist) {
      return { error: 'User already exist' }
    }

    const bcrypt = await bcryptjs.hash(password, 8)
    const userId = await this.mongoService.create<User>('users', {
      id: '',
      createdAt: new Date(),
      services: {
        password: {
          bcrypt
        }
      },
      emails: [
        { email, verified: true }
      ],
      profile: {
        email: email,
        name: email.substr(0, email.indexOf('@'))
      }
    })

    const user = await this.mongoService.get<User>('users', userId)
    if (user) {
      return { user }
    } else {
      return { error: `Error getting user with ID [${userId}]` }
    }
  }

  @Resolver({ type: 'mutation' })
  async loginWithEmailPassword (input: {
    email: string, password: string
  }): Promise<{ error?: string, userId?: string, accessToken?: string, refreshToken?: string }> {
    const { email, password } = input
    const filter = {
      emails: {
        email,
        verified: true
      }
    }
    const user = await this.mongoService.get<User>('users', filter)
    if (user) {
      const { services } = user
      const userBcrypt = services && services.password && services.password.bcrypt
      if (userBcrypt && await bcryptjs.compare(password, userBcrypt)) {
        // login successful, return tokens with user ID as payload
        const payload = {
          userId: user.id
        }
        return {
          userId: user.id,
          ...this.generateTokens(payload)
        }
      }
    }
    return { error: 'Invalid login attempt' }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'query', paramNames: [ 'id' ] })
  async user (id: string): Promise<{ error?: string, user?: User }> {
    const user = await this.mongoService.get<User>('users', id)
    if (!user) {
      return { error: `Error getting user [${id}]` }
    } else {
      return { user }
    }
  }

  /**
   * Create test/seed users within the system.
   *
   * Resolver authorization prevents execution of this function unless environment variable 'seed' is set to '1'.
   */
  @Resolver({ auth: environmentVariableIsSet('seed'), type: 'mutation' })
  async seedUsers (input: { force?: boolean, userCount?: number } = {}): Promise<{ error?: string }> {
    const { force = false, userCount = 20 } = input
    if (!force) {
      // prevent seeding if users exist
      const Users = await this.mongoService.collection('users')
      const userCount = await Users.count({})

      if (userCount > 0) {
        return { error: 'Cannot seed database with users' }
      }
    } else {
      // remove seeded users
      await this.mongoService.remove('users', {
        $or: [
          { 'emails.email': { $regex: 'user[0-9]+@test\.local' } },
          { 'emails.email': 'admin@test.local' }
        ]
      })
    }

    let signUp = await this.signUp({ email: 'admin@test.local', password: 'testAdmin' })
    if (signUp.error) {
      return signUp
    }

    await this.roleService.addUsersToRoles({
      userIds: signUp!.user!.id,
      roles: DefaultRoles.Administrator
    })

    const userIds = []
    for (let c = 1; c <= userCount; c++) {
      let signUp = await this.signUp({ email: `user${c}@test.local`, password: 'testUser' })
      if (signUp.error) {
        return signUp
      }
      userIds.push(signUp!.user!.id)
    }
    await this.roleService.addUsersToRoles({ userIds, roles: DefaultRoles.User })

    return {}
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'query' })
  users (input: { q?: string, limit?: number, cursor?: string } = {}): Promise<FindWithCursor<User>> {
    const { q, limit = 10, cursor } = input
    let filter = {}
    if (q && q.trim().length > 0) {
      filter = {
        $or: [
          { 'profile.email': { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { 'profile.email': { $regex: `${escapeRegex(q)}$`, $options: 'i' } },

          { 'profile.name': { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { 'profile.name': { $regex: `${escapeRegex(q)}$`, $options: 'i' } }
        ]
      }
    }

    return this.mongoService.findWithCursor<User>('users', filter, limit, cursor)
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation', paramNames: [ 'id' ] })
  async removeUser (id: String) {
    return this.mongoService.remove('users', id)
  }

  @Resolver({ auth: userIdMatchParam('id'), type: 'mutation' })
  async updateProfile (input: { id: string, profile: { [key: string]: any } }): Promise<{ error?: string }> {
    const { id, profile } = input
    const profileWithPrefix = Object.keys(profile).reduce((p: { [key: string]: any }, key: string) => {
      p[ `profile.${key}` ] = profile[ key ]
      return p
    }, {})

    await this.mongoService.update('users', id, {
      $set: {
        ...profileWithPrefix
      }
    })

    return {}
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async reset (): Promise<{ error?: string }> {
    const users = await this.mongoService.findWithCursor<User>('users', {
      'emails.email': 'admin@test.local'
    })

    if (users.docs.length === 0) {
      return { error: 'Pre-requisite to reset not met' }
    }

    return this.mongoService.dropCollection('users')
  }

  // --- stuff below... TODO: not done...

  @Resolver({ auth: userIdMatchParam('id'), type: 'mutation' })
  async setProfile<T extends Profile> (input: { id: string, profile: T }): Promise<void> {
    const { id, profile } = input
    return this.mongoService.update('users', id, { profile })
  }

  async getProfile<T extends Profile> (input: { id: string }): Promise<T> {
    const { id } = input
    const user = await this.mongoService.get<User>('users', id, {
      fields: { profile: 1 }
    })
    if (!user) {
      throw new Error(`Error getting profile for user [${id}]`)
    }

    return user.profile as T
  }

  async loginWithUserId (id: string): Promise<{ error?: string, user?: User }> {
    const user = await this.mongoService.get<User>('users', id)
    if (user) {
      return { user }
    }
    return { error: 'Invalid login attempt' }
  }

  verifyToken (token: string): any {
    return jwt.verify(token, this.secretOrKey)
  }

  private generateTokens (payload: any) {
    const accessToken = jwt.sign(payload, this.secretOrKey, {
      expiresIn: this.accessTokenExpiresIn
    })
    // TODO use var randtoken = require('rand-token')
    const refreshToken = jwt.sign(payload, this.secretOrKey, {
      expiresIn: this.refreshTokenExpiresIn
    })

    return { accessToken, refreshToken }
  }
}
