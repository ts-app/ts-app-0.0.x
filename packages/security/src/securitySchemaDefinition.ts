import { MongoService } from '@ts-app/mongo'
import { SchemaDefinition, standardSchemaDefinition, ResolverService } from '@ts-app/graphql'
import { MongoSecurityService } from './MongoSecurityService'
import { MongoRoleService } from './MongoRoleService'
import { loadFile, User } from '@ts-app/common'
import { mergeTypeDefs, ResolverError } from '@ts-app/graphql'

/**
 * Returns a SchemaDefinition object that provides support for access authorization and user
 * account management.
 *
 * The schema definition contains:
 *
 * - role.graphql from this package
 * - security.graphql from this package
 * - defines standardSchemaDefinition from @ts-app/graphql as a dependency
 *
 * @param {MongoService} mongoService
 * @return {SchemaDefinition}
 */
export const securitySchemaDefinition = ({ mongoService }: { mongoService: MongoService }): SchemaDefinition => {
  const resolver = ResolverService.getInstance()
  const roleService = new MongoRoleService({ mongoService })
  resolver.registerService(roleService)
  resolver.registerService(new MongoSecurityService({ mongoService, roleService }))

  // perform authorization checks @Resolver({auth}) is a "function"
  resolver.registerBeforeware(async (input, next, prev) => {
    if (typeof input.metadata.auth === 'function') {
      const auth = input.metadata.auth

      if (auth(input)) {
        next(prev)
      } else {
        throw new ResolverError('Unauthorized access!')
      }
    } else {
      next(prev)
    }
  })

  // perform authorization check if @Resolver({auth}) is a string array
  resolver.registerBeforeware(async (input, next, prev) => {
    const roles = input.metadata.auth
    const user: User = input.context.user

    if (Array.isArray(roles) && roles.length > 0) {
      if (!user) {
        throw new ResolverError('Unauthenticated access!')
      }

      const authorized = await
        roleService.userIsInRoles({
          userId: user.id,
          roles
        })
      if (authorized) {
        // authorized to access restricted resource
        next(prev)
      } else {
        throw new ResolverError('Unauthorized access!')
      }
    } else {
      // will not check if metadata.auth is no an array (of string)
      next(prev)
    }
  })

  const typeDefs = mergeTypeDefs([
    loadFile(`${__dirname}/role.graphqls`),
    loadFile(`${__dirname}/security.graphqls`) ])

  const resolvers = resolver.makeResolvers()
  const standard = standardSchemaDefinition()

  return {
    typeDefs,
    resolvers,
    dependencies: {
      standard
    }
  }
}
