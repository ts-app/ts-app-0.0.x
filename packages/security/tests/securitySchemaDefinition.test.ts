import { ApolloFetch, createApolloFetch } from 'apollo-fetch'
import {
  mergeSchemaDefinitions,
  SchemaDefinition,
  ResolverService,
  serveSchema
} from '@ts-app/graphql'
import { Server } from 'http'
import gql from 'graphql-tag'
import { MongoService } from '@ts-app/mongo'
import { securitySchemaDefinition } from '../src/securitySchemaDefinition'
import { makeExecutableSchema } from 'graphql-tools'
import { MongoSecurityService } from '../src/MongoSecurityService'

describe('securitySchemaDefinition', async () => {
  const localUrl = 'mongodb://localhost:27017'
  let security: SchemaDefinition
  let mongoService: MongoService
  const resolverService = ResolverService.getInstance()
  let server: Server
  let fetch: ApolloFetch

  beforeAll(async () => {
    mongoService = new MongoService(localUrl)
    security = securitySchemaDefinition({ mongoService })
    const merged = mergeSchemaDefinitions([ security ])
    server = await serveSchema({ schema: makeExecutableSchema(merged) })
    fetch = createApolloFetch({ uri: 'http://localhost:3000/graphql' })
  })

  afterAll(async () => {
    // shutdown graphql
    await new Promise(resolve => server.close(resolve))
    resolverService.resetServices()

    // shutdown services
    const db = await mongoService.db()
    await db.close()
  })

  beforeEach(async () => {
    try {
      await mongoService.dropCollection('users')
      await mongoService.dropCollection('roles')
    } catch {
      // it's ok
    }
  })

  afterEach(() => {
    resolverService.setLogger({ error: console.error })
  })

  const hideResolverServiceLog = () => {
    resolverService.setLogger({
      error: () => {
        // do nothing
      }
    })
  }

  const createFetchAs = (accessToken: string) => {
    const fetch = createApolloFetch({ uri: 'http://localhost:3000/graphql' })
    fetch.use(({ request, options }, next) => {
      if (!options.headers) {
        options.headers = {}
      }
      options.headers[ 'Authorization' ] = 'Bearer ' + accessToken
      next()
    })
    return fetch
  }

  const querySignUp = gql`mutation($email: String!, $password: String!){
      signUp(email: $email, password: $password) {
          error
          user {
              id
              createdAt
              emails { email verified }
          }
      }
  }`

  test('security schema definition', async () => {
    expect(security).toMatchSnapshot()
  })

  test('signUp() with invalid email', async () => {
    try {
      const result = await fetch({
        query: querySignUp, variables: { email: 'bob@bob', password: '123' }
      })
      expect(result).toMatchSnapshot()
    } catch (e) {
      console.error(e)
    }

  })

  test('signUp() without password', async () => {
    const result = await fetch({
      query: querySignUp, variables: { email: 'bob@bob.com', password: '' }
    })
    expect(result).toMatchSnapshot()
  })

  test('signUp() works', async () => {
    const result = await fetch({
      query: querySignUp, variables: { email: 'bob@bob.com', password: '123' }
    })
    expect(result.errors).toBeFalsy()
    expect(result.data.signUp.user.id.length).toBe(24)
  })

  test('loginWithEmailPassword works', async () => {
    let result = await fetch({
      query: querySignUp, variables: { email: 'bob@bob.com', password: '123' }
    })
    expect(result.errors).toBeFalsy()
    const signUp = result.data.signUp

    result = await fetch({
      query: gql`mutation {
          loginWithEmailPassword(email: "bob@bob.com", password: "123") {
              error accessToken refreshToken
          }
      }`
    })

    // decode token & verify userId
    const { accessToken } = result.data.loginWithEmailPassword
    const securityService = resolverService.getService<MongoSecurityService>(MongoSecurityService)!
    const token = securityService.verifyToken(accessToken)
    expect(token.userId).toBe(signUp.user.id)
  })

  test('allow anonymous access to "signUp"', async () => {
    const result = await fetch({
      query: gql`mutation {
          signUp(email: "user@test.local", password: "secret") {
              error
              user {
                  id
              }
          }
      }`
    })

    expect(result.data.signUp.error).toBeFalsy()
    expect(result.data.signUp.user.id.length).toBe(24)
  })

  test('allow anonymous access to "loginWithEmailPassword"', async () => {
    const result = await fetch({
      query: gql`mutation {
          loginWithEmailPassword(email: "someone", password: "someone") {
              error accessToken refreshToken
          }
      }`
    })

    expect(result.data).toMatchSnapshot()
  })

  test('prevent anonymous access to "users"', async () => {
    hideResolverServiceLog()

    const result = await fetch({
      query: gql`query {
          users {
              error
              docs {
                  id
              }
          }
      }`
    })

    expect(result.data.users.error).toBeTruthy()
    expect(result.data.users).toMatchSnapshot()
  })

  test('authorized access to "users"', async () => {
    hideResolverServiceLog()

    const securityService = resolverService.getService('SecurityService') as MongoSecurityService
    const seed = await securityService.seedUsers()
    expect(seed.error).toBeFalsy()

    // anonymous can login as user1
    const user1 = await fetch({
      query: gql`mutation {
          loginWithEmailPassword(email: "user1@test.local", password: "testUser") {
              error
              accessToken
          }
      }`
    })
    const user1AccessToken = user1.data.loginWithEmailPassword.accessToken
    expect(user1AccessToken.length).toBeGreaterThan(0)

    // user1 cannot access "users"
    const userQuery = {
      query: gql`query {
          users {
              error
              docs {
                  profile {
                      email
                  }
              }
          }
      }`
    }
    const fetchAsUser1 = createFetchAs(user1AccessToken)
    let result = await fetchAsUser1(userQuery)
    let { users } = result.data
    expect(users.error).toBe('Unauthorized access!')
    expect(users.docs).toBeFalsy()

    const admin = await fetch({
      query: gql`mutation {
          loginWithEmailPassword(email: "admin@test.local", password:"testAdmin") {
              error accessToken
          }
      }`
    })

    const fetchAsAdmin = createFetchAs(admin.data.loginWithEmailPassword.accessToken)
    result = await fetchAsAdmin(userQuery)
    users = result.data.users
    expect(users).toMatchSnapshot()
  })

  test('authorized access with userIdMatchParam()', async () => {
    hideResolverServiceLog()

    const securityService = resolverService.getService('SecurityService') as MongoSecurityService
    const seed = await securityService.seedUsers()
    expect(seed.error).toBeFalsy()

    // anonymous can login as user1
    const user1 = await fetch({
      query: gql`mutation {
          loginWithEmailPassword(email: "user1@test.local", password: "testUser") {
              error
              accessToken
              userId
          }
      }`
    })
    const user1AccessToken = user1.data.loginWithEmailPassword.accessToken
    expect(user1AccessToken.length).toBeGreaterThan(0)

    // user1 can update profile for user1
    let updateProfile = {
      query: gql`mutation {
          updateProfile(id: "${user1.data.loginWithEmailPassword.userId}", profile: {name: "bob"}) {
              error
          }
      }`
    }
    const fetchAsUser1 = createFetchAs(user1AccessToken)
    let result = await fetchAsUser1(updateProfile)
    expect(result.data.updateProfile.error).toBeFalsy()

    // --- negative test case
    // anonymous can login as user2
    const user2 = await fetch({
      query: gql`mutation {
          loginWithEmailPassword(email: "user2@test.local", password: "testUser") {
              error
              accessToken
              userId
          }
      }`
    })
    const user2AccessToken = user2.data.loginWithEmailPassword.accessToken
    expect(user2AccessToken.length).toBeGreaterThan(0)

    // user2 CANNOT update profile for user1
    updateProfile = {
      query: gql`mutation {
          updateProfile(id: "${user1.data.loginWithEmailPassword.userId}", profile: {name: "bob"}) {
              error
          }
      }`
    }
    const fetchAsUser2 = createFetchAs(user2AccessToken)
    result = await fetchAsUser2(updateProfile)
    expect(result).toMatchSnapshot()
  })

})
