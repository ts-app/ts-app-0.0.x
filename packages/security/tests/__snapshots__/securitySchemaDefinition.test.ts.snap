// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`securitySchemaDefinition allow anonymous access to "loginWithEmailPassword" 1`] = `
Object {
  "loginWithEmailPassword": Object {
    "accessToken": null,
    "error": "Invalid login attempt",
    "refreshToken": null,
  },
}
`;

exports[`securitySchemaDefinition authorized access to "users" 1`] = `
Object {
  "docs": Array [
    Object {
      "profile": Object {
        "email": "admin@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user1@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user2@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user3@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user4@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user5@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user6@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user7@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user8@test.local",
      },
    },
    Object {
      "profile": Object {
        "email": "user9@test.local",
      },
    },
  ],
  "error": null,
}
`;

exports[`securitySchemaDefinition authorized access with userIdMatchParam() 1`] = `
Object {
  "data": Object {
    "updateProfile": Object {
      "error": "Unauthorized access!",
    },
  },
}
`;

exports[`securitySchemaDefinition prevent anonymous access to "users" 1`] = `
Object {
  "docs": null,
  "error": "Unauthenticated access!",
}
`;

exports[`securitySchemaDefinition security schema definition 1`] = `
Object {
  "dependencies": Object {
    "standard": Object {
      "resolvers": Object {
        "Date": "Date",
      },
      "typeDefs": "scalar Date",
    },
  },
  "resolvers": Object {
    "Mutation": Object {
      "addUsersToRoles": [Function],
      "createRole": [Function],
      "loginWithEmailPassword": [Function],
      "removeRole": [Function],
      "removeUser": [Function],
      "removeUsersFromAllRoles": [Function],
      "removeUsersFromRoles": [Function],
      "reset": [Function],
      "seedUsers": [Function],
      "setProfile": [Function],
      "signUp": [Function],
      "updateProfile": [Function],
    },
    "Query": Object {
      "findRolesWithCursor": [Function],
      "getGroupsForUser": [Function],
      "getUsersInRoles": [Function],
      "user": [Function],
      "userIsInRoles": [Function],
      "userRoles": [Function],
      "users": [Function],
    },
  },
  "typeDefs": "type Query {
    userIsInRoles(userId: String!, roles: [String]!, group: String): Boolean
    getGroupsForUser(userId: String!, role: String): GetGroupsForUserPayload
    findRolesWithCursor(q: String, limit: Int, cursor: String): FindRolesWithCursorPayload
    userRoles(userId: String!, group: String): GetRolesForUserPayload
    getUsersInRoles(roles: [String]!, group: String, limit: Int, cursor: String): GetUsersInRolesPayload
user(id: ID!): UserPayload
    users(q: String, limit: Int, cursor: String): UsersPayload
}

type Mutation {
    addUsersToRoles(userIds: [String]!, roles: [String], group: String): NoPayload
    createRole(name: String!): CreateRolePayload
    removeRole(name: String!): NoPayload
    removeUsersFromRoles(userIds: [String]!, roles: [String]!, group: String): NoPayload
    removeUsersFromAllRoles(userIds: [String]!): NoPayload
seedUsers(force: Boolean): NoPayload
    reset: NoPayload
    signUp(email: String!, password: String!): SignUpPayload
    loginWithEmailPassword(email: String!, password: String!): LoginPayload
    removeUser(id: ID!): NoPayload
    updateProfile(id: ID!, profile: ProfileInput!): NoPayload
    setProfile(id:ID!, profile: ProfileInput!): NoPayload
    updateUser(id: ID!, user: UserInput!): UserPayload
}

type Role {
    id: String
    name: String
}

type CreateRolePayload {
    error: String
    id: String
}

type GetGroupsForUserPayload {
    error: String
    groups: [String]
}

type FindRolesWithCursorPayload {
    error: String
    cursor: String
    docs: [Role]
}

type GetRolesForUserPayload {
    error: String
    roles: [String]
}

type GetUsersInRolesPayload {
    error: String
    cursor: String
    docs: [User]
}





input UserInput {
    emails: [EmailInput]
    services: [UserServiceInput]
    roles: [RoleGroupInput]
    profile: ProfileInput
}

input RoleGroupInput {
    role: String
    group: String
}

input EmailInput {
    email: String
    verified: Boolean
}

input UserServiceInput {
    password: ServicePasswordInput
}

input ServicePasswordInput {
    bcrypt: String!
}

input ServiceJwtInput {
    refreshTokens: [JwtRefreshTokenInput]
}

input JwtRefreshTokenInput {
    name: String!
    token: String!
}

input ProfileInput {
    email: String
    name: String
    avatarUrl: String
}

type User {
    id: ID!
    createdAt: Date
    emails: [Email]
    services: [UserService]
    roles: [RoleGroup]
    profile: Profile
}

type RoleGroup {
    role: String
    group: String
}

type Email {
    email: String
    verified: Boolean
}

type UserService {
    password: ServicePassword
}

type ServicePassword {
    bcrypt: String!
}

type ServiceJwt {
    refreshTokens: [JwtRefreshToken]
}

type JwtRefreshToken {
    name: String!
    token: String!
}

type Profile {
    email: String!
    name: String!
    avatarUrl: String
}

type LoginPayload {
    error: String
    userId: String
    accessToken: String
    refreshToken: String
}

type SignUpPayload {
    error: String
    user: User
}

type UsersPayload {
    error: String
    cursor: String
    docs: [User]
}

type UserPayload {
    error: String
    user: User
}

type NoPayload {
    error: String
}
",
}
`;

exports[`securitySchemaDefinition signUp() with invalid email 1`] = `
Object {
  "data": Object {
    "signUp": Object {
      "error": "Invalid email",
      "user": null,
    },
  },
}
`;

exports[`securitySchemaDefinition signUp() without password 1`] = `
Object {
  "data": Object {
    "signUp": Object {
      "error": "Invalid password",
      "user": null,
    },
  },
}
`;
